numpy/core/src/multiarray/item_selection.c

```diff
@@ -1750,6 +1751,275 @@ PyArray_LexSort(PyObject *sort_keys, int axis)
+            || !(PyArray_FLAGS(mps[j]) & NPY_ARRAY_ALIGNED)
+            || (PyArray_STRIDES(mps[j])[axis] != (npy_intp)PyArray_DESCR(mps[j])->elsize);
+        maxelsize = PyArray_MAX(maxelsize, PyArray_DESCR(mps[j])->elsize);
+    }
+    assert(N > 0);  /* Guaranteed and assumed by indbuffer */
+    npy_intp *indbuffer = PyDataMem_NEW(N * sizeof(*indbuffer));
+    if (indbuffer == NULL) {
+       PyErr_NoMemory();
+       goto fail;
+    }
+    if (needcopy) {
+        char *valbuffer;
+        // Create a buffer of size of the current 1D slice i.e N
+        npy_intp valbufsize = N * maxelsize;
+        if (NPY_UNLIKELY(valbufsize) == 0) {
+            valbufsize = 1;  /* Ensure allocation is not empty */
+        }
+        valbuffer = PyDataMem_NEW(valbufsize);
+        if (valbuffer == NULL) {
+            PyDataMem_FREE(indbuffer);
+            PyErr_NoMemory();
+            goto fail;
+        }
+        while (size--) {
+            for (i = 0; i < N; i++) {
+                indbuffer[i] = i;
+            }
+            for (j = 0; j < n; j++) {
+                int rcode;
+                elsize = PyArray_DESCR(mps[j])->elsize;
+                astride = PyArray_STRIDES(mps[j])[axis];
+                argsort = PyArray_DESCR(mps[j])->f->argsort[NPY_STABLESORT];
+                if(argsort == NULL) {
+                    argsort = npy_atimsort;
+                }
+                _unaligned_strided_byte_copy(valbuffer, (npy_intp) elsize,
+                                             its[j]->dataptr, astride, N, elsize);
+                if (PyArray_ISBYTESWAPPED(mps[j])) {
+                    _strided_byte_swap(valbuffer, (npy_intp) elsize, N, elsize);
+                }
+                rcode = argsort(valbuffer, indbuffer, N, mps[j]);
+                if (rcode < 0 || (PyDataType_REFCHK(PyArray_DESCR(mps[j]))
+                            && PyErr_Occurred())) {
+                    PyDataMem_FREE(valbuffer);
+                    PyDataMem_FREE(indbuffer);
+                    goto fail;
+                }
+                PyArray_ITER_NEXT(its[j]);
+            }
+            // For unaligned we use stride to copy
+            for(npy_intp i = 0; i < N; i++){
+                char *current = PyDataMem_NEW(self_elsize);
+                if (current == NULL){
+                   PyDataMem_FREE(indbuffer);
+                   PyErr_NoMemory();
+                   goto fail;
+                }
+                memmove(current, sit->dataptr + (i*self_stride), self_elsize);
+                npy_intp current_idx = i;
+                while(1){
+                    npy_intp next_idx = indbuffer[current_idx];
+                    indbuffer[current_idx] = current_idx;
+                    if (next_idx == i) {
+                        break;
+                    }
+                    memmove(sit->dataptr +(current_idx*self_stride), sit->dataptr + (next_idx*self_stride), self_elsize);
+                    current_idx = next_idx;
```

Comment:
I think it would be good to refactor this into a small helper. Deduplicates the code/explains what is going on in this case (through the function name), and adds an obvious place to add a comment of whats going on, and maybe a reference for the approach (if you have).

I will look at the rest soon, sorry, please ping me if I forget...

Reply:
Sure, will refactor it into a helper.